presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,397 @@
+        };
+    }
+
+    private static class Visitor
+            extends io.prestosql.sql.tree.ExpressionRewriter<Void>
+    {
+        private final Metadata metadata;
+        private final TypeAnalyzer typeAnalyzer;
+        private final Session session;
+        private final TypeProvider types;
+        private final InterpretedFunctionInvoker functionInvoker;
+        private final LiteralEncoder literalEncoder;
+
+        public Visitor(Metadata metadata, TypeAnalyzer typeAnalyzer, Session session, TypeProvider types)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+            this.session = requireNonNull(session, "session is null");
+            this.types = requireNonNull(types, "types is null");
+            this.functionInvoker = new InterpretedFunctionInvoker(metadata.getFunctionRegistry());
+            this.literalEncoder = new LiteralEncoder(metadata.getBlockEncodingSerde());
+        }
+
+        @Override
+        public Expression rewriteComparisonExpression(ComparisonExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            ComparisonExpression expression = (ComparisonExpression) treeRewriter.defaultRewrite((Expression) node, null);
+            return unwrapCast(expression);
+        }
+
+        private Expression unwrapCast(ComparisonExpression expression)
+        {
+            if (!(expression.getLeft() instanceof Cast)) {
+                return expression;
+            }
+
+            Object right = ExpressionInterpreter.expressionOptimizer(expression.getRight(), metadata, session, typeAnalyzer.getTypes(session, types, expression.getRight()))
+                    .optimize(NoOpSymbolResolver.INSTANCE);
+
+            if (right instanceof Expression) {
+                return expression;
+            }
+
+            Cast cast = (Cast) expression.getLeft();
+            ComparisonExpression.Operator operator = expression.getOperator();
+
+            if (right == null) {
+                switch (operator) {
+                    case EQUAL:
+                    case NOT_EQUAL:
+                    case LESS_THAN:
+                    case LESS_THAN_OR_EQUAL:
+                    case GREATER_THAN:
+                    case GREATER_THAN_OR_EQUAL:
+                        return new Cast(new NullLiteral(), BooleanType.BOOLEAN.toString());
+                    case IS_DISTINCT_FROM:
+                        return TRUE_LITERAL;
+                    default:
+                        throw new UnsupportedOperationException("Not yet implemented");
+                }
+            }
+
+            Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
+            Type targetType = typeAnalyzer.getType(session, types, expression.getRight());
+
+            if (!isInjectiveConversion(sourceType, targetType)) {
```

Comment:
Obviously the code below relies on the fact that implicit coercion is monotonic.
However, the code as it is now no longer explicitly check it's a coercion. Now the method doing the check is called "isInjectiveConversion" (so it could return true for `bigint → varchar(19)` too). Please rename the method to convey the meaning.
`isInjectiveCoercion` might be OK, although doesn't signal this really should be "implicit coercion"

Reply:
`isInjectiveMonotonicConversion`? `isStrictlyMonotonic..`?

Reply:
I see how it might be confusing. I'll change it to `hasInjectiveImplicitCoercion`